### ISSUE TITLE
cloudpathlib type checking doesn't work well with pyright (but works ok with mypy)

### ISSUE THREAD

**Issue by karolzlot** [microsoft/pyright#3117]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

**Describe the bug**
Type checking doesn't work well for [cloudpathlib](https://github.com/drivendataorg/cloudpathlib) objects if using pyright (but works well with mypy)

**To Reproduce**
Steps to reproduce the behavior.

Create instance of `S3Path` class. Then you can see that `instance.name` should be detected as `str` type by pyright, but pyright isn't detecting it as `str` type.

**Expected behavior**
Property `.name` should be detected by pyright as `str` (many other properties in this library have the same problem)

**Screenshots or Code**
```python
from cloudpathlib import S3Path
instance1= S3Path("s3://example/data/")
name1:str = instance1.name
```

![image](https://user-images.githubusercontent.com/11590960/155880831-712edb13-032c-42f4-8d3f-34fcb6bbe830.png)

---
![image](https://user-images.githubusercontent.com/11590960/155880818-d35eea7a-e70f-4230-9689-cd169c2c421d.png)

**VS Code extension or command-line**
Are you running pyright as a VS Code extension or a command-line tool? **Yes**
Which version? **v1.1.225**

**Additional context**

I first reported it here:
https://github.com/drivendataorg/cloudpathlib/issues/179
but now I think it should be fixed in pyright (but I'm not sure...)

**Comment by hmc-cs-mdrissi** [comment]
Your pyright config is relevant here. I can reproduce this with` "useLibraryCodeForTypes": true `but not with `"useLibraryCodeForTypes": false`. First, cloudpathlib is not a py.typed library so it's treated as not having type information. If a library that uses type annotations wants to have that information be used it needs to mark itself appropriately as mentioned [here](https://www.python.org/dev/peps/pep-0561/#packaging-type-information). In practice add an empty file called py.typed at top folder for package.

Because library is untyped pyright infers types if you enable useLibraryCodeForTypes: true (default in pylance). Now why is .name inferred to not be a string. Well code for it https://github.com/drivendataorg/cloudpathlib/blob/master/cloudpathlib/cloudpath.py#L541 does not infer a type of string. self._dispatch_to_path has a broader return type. If you try adding a type annotation of str to the name property it will give a proper error of _dispatch_to_path may return other types.

So overall this doesn't look like a bug here and mainly an issue with missing types/type errors in cloudpath library. The code near bottom of _dispatch_to_path even explicitly mentions,

`cases this should include: str, empty sequence, sequence of str, ...`

as some return types. My recommendations fix type errors in it, include missing types for public interface (pyright --verifytypes can help here) and then add py.typed file to it.

**Comment by erictraut** [comment]
As @hmc-cs-mdrissi said, this is an untyped library. By default, pyright treats all symbols exported from an untyped library as `Unknown`. This is consistent with the behavior of mypy. I presume that you have enabled the `useLibraryCodeForTypes` setting. When this is enabled, pyright attempts to infer types based on the library's source code. This feature is not generally recommended for static type checking. It can evaluate types that are good enough for completion suggestions, but inferred types are not generally accurate or complete enough for static type checking and will often result in false positive errors. 

If you want to use this library with static type checking, you should look for type stubs (".pyi" files) that describe the types for the library. Or talk to the library's maintainers about providing inlined type information in the form of type annotations. [This guidance](https://github.com/microsoft/pyright/blob/main/docs/typed-libraries.md) may be of help to them.

In any case, this is not a bug in pyright, so I'm going to close the issue.

**Comment by karolzlot** [comment]
Yes, I had `"useLibraryCodeForTypes": true` set: 

https://github.com/drivendataorg/cloudpathlib/issues/179#issuecomment-1054851493
